docs/proposals/network-policy.md

```diff
@@ -0,0 +1,316 @@
+	Isolation *IngressIsolationPolicy `json:"isolation,omitempty"` 
+}
+```
+
+```yaml
+kind: Namespace
+apiVersion: v1
+spec:
+  networkPolicy:
+    ingress:
+      isolation: DefaultDeny
+```
+
+The above structures will be represented in v1beta1 as a json encoded annotation like so:
+
+```yaml
+kind: Namespace
+apiVersion: v1
+metadata:
+  annotations:
+    net.beta.kubernetes.io/network-policy: "{\"ingress\": {\"isolation\": \"DefaultDeny\"}}" 
+```
+
+### NetworkPolicy Go Definition
+
+Once a Namespace is isolated, a mechanism to selectively allow traffic into the namespace and between pods within
+the namespace is required.  That is accomplished through ingress rules on `NetworkPolicy`
+objects (of which there can be multiple in a single namespace).  Pods selected by
+one or more NetworkPolicy objects should allow any incoming connections that match any
+ingress rule on those NetworkPolicy objects, per the network plugin’s capabilities.
+
+If ingress isolation is not specified on a namespace, then all traffic is allowed to pods in that namespace.
+
+NetworkPolicy objects and the above namespace isolation both act on _connections_ rather than packets.  That is to say that if traffic from pod A to pod B is allowed by the configured
+policy, then the return packets for that connection from B -> A are also allowed, even if policy is in place that would prevent 
+B from initiating a connection to A.  If new network policy is applied that would block an existing connection between two endpoints, the enforcer of policy 
+should terminate and block the existing connection as soon as can be expected by the implementation.
+
+We propose adding a new `/apis/networking` API group for the NetworkPolicy object.
+
+```go
+type NetworkPolicy struct {
+  TypeMeta
+  ObjectMeta
+
+  // Specification of the desired behavior for this NetworkPolicy.
+  Spec NetworkPolicySpec 
+}
+
+type NetworkPolicySpec struct {
+  // Selects the pods to which this NetworkPolicy object applies.  The array of NetworkPolicyIngressRules below 
+  // is applied to any pods selected by this field. Multiple NetworkPolicy objects can select the 
+  // same set of pods.  In this case, the NetworkPolicyRules for each are combined additively.
+  // This field is NOT optional and follows standard unversioned.LabelSelector semantics.  
+  // An empty PodSelector matches all pods in this namespace.
+  PodSelector unversioned.LabelSelector `json:"podSelector"`
+
+  // List of ingress rules to be applied to the selected pods.
+  // Traffic is allowed to a pod if Namespace.NetworkPolicy.Ingress.Isolation is undefined, 
+  // OR if the traffic source is the pod's local kubelet (for health checks), 
+  // OR if the traffic matches at least one NetworkPolicyIngressRule across all of the NetworkPolicy 
+  // objects whose podSelector matches the pod.  
+  // If this field is nil, this NetworkPolicy does not affect ingress to the selected pods.
```

Comment:
@MikeSpreitzer - The user facing API documentation is auto-generated from the versioned API Go code comments.  They are one and the same.  We'll need a NetworkPolicy user-guide, but that is separate from the API documentation.

I think your view of how we got here is not right - this doc is not "in response" to any of the previous design docs that were written regarding network policy - it takes the content of what we discussed in the other docs, calls, etc, but is modeled after other Kubernetes API change proposals.  

See:
- https://github.com/kubernetes/kubernetes/blob/master/docs/proposals/pod-security-context.md
- https://github.com/kubernetes/kubernetes/blob/master/docs/proposals/deployment.md

> Casey wanted to write in psuedo-code instead, and needed to declare the data structures processed by that psuedo-code

You had originally asked me to write the pseudo-code because the english was not clear enough.  That came well after the Go structs were defined in this proposal (see the list of commits on this proposal, commit `3bbd0e6cf0fd`).